internal/ls/changetrackerimpl.go

```diff
@@ -0,0 +1,413 @@
+				// assert change[i].End <= change[i + 1].Start
+				panic(fmt.Sprintf("changes overlap: %v and %v", changesInFile[i].Range, changesInFile[i+1].Range))
+			}
+		}
+
+		textChanges := core.MapNonNil(changesInFile, func(change *trackerEdit) *lsproto.TextEdit {
+			// !!! targetSourceFile
+
+			newText := ct.computeNewText(change, sourceFile, sourceFile)
+			// span := createTextSpanFromRange(c.Range)
+			// !!!
+			// Filter out redundant changes.
+			// if (span.length == newText.length && stringContainsAt(targetSourceFile.text, newText, span.start)) { return nil }
+
+			return &lsproto.TextEdit{
+				NewText: newText,
+				Range:   change.Range,
+			}
+		})
+
+		if len(textChanges) > 0 {
+			changes[sourceFile.FileName()] = textChanges
+		}
+	}
+	return changes
+}
+
+func (ct *changeTracker) computeNewText(change *trackerEdit, targetSourceFile *ast.SourceFile, sourceFile *ast.SourceFile) string {
+	switch change.kind {
+	case trackerEditKindRemove:
+		return ""
+	case trackerEditKindText:
+		return change.NewText
+	}
+
+	pos := int(ct.ls.converters.LineAndCharacterToPosition(sourceFile, change.Range.Start))
+	formatNode := func(n *ast.Node) string {
+		return ct.getFormattedTextOfNode(n, targetSourceFile, sourceFile, pos, change.options)
+	}
+
+	var text string
+	switch change.kind {
+
+	case trackerEditKindReplaceWithMultipleNodes:
+		if change.options.joiner == "" {
+			change.options.joiner = ct.newLine
+		}
+		text = strings.Join(core.Map(change.nodes, func(n *ast.Node) string { return strings.TrimSuffix(formatNode(n), ct.newLine) }), change.options.joiner)
+	case trackerEditKindReplaceWithSingleNode:
+		text = formatNode(change.Node)
+	default:
+		panic(fmt.Sprintf("change kind %d should have been handled earlier", change.kind))
+	}
+	// strip initial indentation (spaces or tabs) if text will be inserted in the middle of the line
+	noIndent := text
+	if !(change.options.indentation != nil && *change.options.indentation != 0 || format.GetLineStartPositionForPosition(pos, targetSourceFile) == pos) {
+		noIndent = strings.TrimLeftFunc(text, unicode.IsSpace)
+	}
+	return change.options.prefix + noIndent // !!!  +((!options.suffix || endsWith(noIndent, options.suffix)) ? "" : options.suffix);
+}
+
+/** Note: this may mutate `nodeIn`. */
+func (ct *changeTracker) getFormattedTextOfNode(nodeIn *ast.Node, targetSourceFile *ast.SourceFile, sourceFile *ast.SourceFile, pos int, options changeNodeOptions) string {
+	text, sourceFileLike := ct.getNonformattedText(nodeIn, targetSourceFile)
+	// !!! if (validate) validate(node, text);
+	formatOptions := getFormatCodeSettingsForWriting(ct.formatSettings, targetSourceFile)
+
+	var initialIndentation, delta int
+	if options.indentation == nil {
+		// !!! indentation for position
+		// initialIndentation = format.GetIndentationForPos(pos, sourceFile, formatOptions, options.prefix == ct.newLine || scanner.GetLineStartPositionForPosition(pos, targetFileLineMap) == pos);
```

Comment:
@jakebailey That part of the formatter isn't ported yet (`getSmartIndent`) and involves the `FindFirstNonWhitespaceColumn` functions, so I didn't include it

Reply:
I wouldn't assume anything else is gonna necessitate porting more parts of `getSmartIndent` - they're only nominally "in the formatter" - none of the `FormatX` entrypoints use anything as-yet unported. Anything else is basically just part of the quickfix pipeline that lives elsewhere, for some reason.